.buildkite/e2e/pipeline-gen/main.go

```diff
@@ -0,0 +1,390 @@
+		"k": EnvVarK8sVersion,
+		"s": EnvVarStackVersion,
+		"t": EnvVarTestsMatch,
+	}
+
+	fixed string
+	mixed string
+
+	output string
+
+	rootDir string
+)
+
+func init() {
+	flag.StringVar(&fixed, "f", "", "Fixed variables (comma-separated list)")
+	flag.StringVar(&mixed, "m", "", "Mixed variables (comma-separated list of colon-separated list)")
+	flag.StringVar(&output, "o", "buildkite-pipeline", "Type of output: buildkite-pipeline or envfile")
+	flag.Parse()
+
+	rand.Seed(time.Now().UTC().UnixNano())
+
+	_, filename, _, ok := runtime.Caller(0)
+	if !ok {
+		fmt.Printf("failed to get current path")
+		os.Exit(1)
+	}
+	rootDir = filepath.Join(filepath.Dir(filename), "../../..")
+}
+
+func main() {
+	stat, err := os.Stdin.Stat()
+	handlErr("Failed to read stdin", err)
+
+	var groups []Group
+
+	// no stdin
+	if stat.Mode()&os.ModeCharDevice != 0 {
+
+		fixedEnv, err := stringListToEnv(fixed)
+		handlErr("Failed to read fixed variables", err)
+
+		mixedEnv, err := stringListToEnvs(mixed)
+		handlErr("Failed to read mixed variables", err)
+
+		groups = []Group{{
+			Fixed: fixedEnv,
+			Mixed: mixedEnv,
+		}}
+
+	} else {
+		in, err := io.ReadAll(os.Stdin)
+		handlErr("Failed to read stdin", err)
+		if len(in) == 0 {
+			handlErr("Failed to read stdin", errors.New("nothing on /dev/stdin"))
+		}
+
+		err = yaml.Unmarshal(in, &groups)
+		handlErr("Failed to parse stdin", err)
+	}
+
+	// build a flat list of the tests to run
+	tests := make([]TestsSuiteRun, 0)
+	cleanup := false
+	for i := range groups {
+		if groups[i].Mixed == nil {
+			groups[i].Mixed = []Env{{}}
+		}
+		for j := range groups[i].Mixed {
+			ts, err := newTestsSuite(groups[i].Label, groups[i].Fixed, len(groups[i].Mixed), groups[i].Mixed[j])
+			handlErr("Failed to create tests suite", err)
+
+			tests = append(tests, ts)
+			cleanup = cleanup || ts.Cleanup
+		}
+	}
+
+	if output == "envfile" {
+		if len(tests) > 1 {
+			handlErr("Not supported with output envfile", errors.New("more than 1 test suite to run"))
+			return
+		}
+		for k, v := range tests[0].Env {
+			fmt.Printf("%s=%s\n", k, v)
+		}
+		return
+	}
+
+	tpl, err := template.New("pipeline.yaml").Parse(pipelineTemplate)
+	handlErr("Failed to parse template", err)
+
+	err = tpl.Execute(os.Stdout, map[string]interface{}{
+		"Cleanup": cleanup,
+		"Tests":   tests,
+	})
+	handlErr("Failed to generate pipeline", err)
+}
+
+func stringListToEnv(str string) (Env, error) {
+	if str == "" {
+		return nil, nil
+	}
+	env := Env{}
+	for _, elem := range strings.Split(str, ",") {
+		kv := strings.Split(elem, "=")
+		if len(kv) != 2 {
+			return nil, fmt.Errorf("no environment variable found in format `k=v` for %s", elem)
+		}
+		env[kv[0]] = kv[1]
+	}
+	return env, nil
+}
+
+func stringListToEnvs(str string) ([]Env, error) {
+	if str == "" {
+		return nil, nil
+	}
+	envs := []Env{}
+	mixedsStr := strings.Split(str, ",")
+	for _, mixedStr := range mixedsStr {
+		mixedEnv := Env{}
+		envStr := strings.Split(mixedStr, ":")
```

Comment:
This means that `mixed` specified as flags cannot contain docker images with `:` to separate image from tag?

Reply:
Very good point. At first, I only supported one single env var per combination in `-f` but eventually found it a shame not support multiple variables. Then I wanted to avoid `;`, `&` or `|` that needs double quotes because they are special shell characters.

How about `^`?

`pipeline-gen -f p=gke -m k=1.22^s=8.5.0,k=1.23^s=8.6.0`

Reply:
I think that would be OK and could be read as a logical conjunction operator which kind of fits the purpose.